Signal/src/ViewControllers/MessagesViewController.m

```diff
@@ -2704,48 +3032,125 @@ - (TSInteraction *)interactionAtIndexPath:(NSIndexPath *)indexPath {
-    if (flag) {
-        NSData *audioData = [NSData dataWithContentsOfURL:recorder.url];
-        SignalAttachment *attachment =
-            [SignalAttachment attachmentWithData:audioData dataUTI:(NSString *)kUTTypeMPEG4Audio filename:nil];
-        if (!attachment ||
-            [attachment hasError]) {
-            DDLogWarn(@"%@ %s Invalid attachment: %@.",
-                      self.tag,
-                      __PRETTY_FUNCTION__,
-                      attachment ? [attachment errorName] : @"Missing data");
-            [self showErrorAlertForAttachment:attachment];
-        } else {
-            [self tryToSendAttachmentIfApproved:attachment];
-        }
+    self.audioRecorder = [[AVAudioRecorder alloc] initWithURL:fileURL
+                                                     settings:@{
+                                                         AVFormatIDKey : @(kAudioFormatMPEG4AAC),
+                                                         AVSampleRateKey : @(44100),
+                                                         AVNumberOfChannelsKey : @(2),
+                                                         AVEncoderBitRateKey: @(128 * 1024),
+                                                     }
+                                                        error:&error];
+    if (error) {
+        DDLogError(@"%@ Couldn't create audioRecorder: %@", self.tag, error);
+        [self resetRecordingVoiceMemo];
+        OWSAssert(0);
+        return;
+    }
+
+    self.audioRecorder.meteringEnabled = YES;
+
+    if (![self.audioRecorder prepareToRecord]) {
+        DDLogError(@"%@ audioRecorder couldn't prepareToRecord.", self.tag);
+        [self resetRecordingVoiceMemo];
+        OWSAssert(0);
+        return;
+    }
+
+    if (![self.audioRecorder record]) {
+        DDLogError(@"%@ audioRecorder couldn't record.", self.tag);
+        [self resetRecordingVoiceMemo];
+        OWSAssert(0);
+        return;
+    }
+}
+
+- (void)endRecordingVoiceMemo
+{
+    OWSAssert([NSThread isMainThread]);
+
+    DDLogInfo(@"endRecordingVoiceMemo");
+
+    if (!self.audioRecorder) {
+        DDLogError(@"%@ Missing audioRecorder", self.tag);
+        OWSAssert(0);
+        return;
+    }
+
+    NSTimeInterval currentTime = self.audioRecorder.currentTime;
+
+    [self.audioRecorder stop];
+
+    const NSTimeInterval kMinimumRecordingTimeSeconds = 1.f;
+    if (currentTime < kMinimumRecordingTimeSeconds) {
+        DDLogInfo(@"Discarding voice memo; too short.");
+        self.audioRecorder = nil;
+        return;
+    }
+
+    NSData *audioData = [NSData dataWithContentsOfURL:self.audioRecorder.url];
+
+    if (!audioData) {
+        DDLogError(@"%@ Couldn't load audioRecorder data", self.tag);
+        OWSAssert(0);
+        self.audioRecorder = nil;
+        return;
+    }
+
+    self.audioRecorder = nil;
+
+    SignalAttachment *attachment =
+        [SignalAttachment attachmentWithData:audioData dataUTI:(NSString *)kUTTypeMPEG4Audio filename:nil];
```

Comment:
Take it or leave it: Until we have a specific "voice note" interface, maybe we ought to have a filename like "Voice Note.m4a" or "Voice Note-[date-stamp].m4a" so that recipients can distinguish it from other audio attachments.